### PR TITLE
Created wizard for redeeming gift cards. #6722

### PR DESCRIPTION
**Pull request by PritishC**

### CODE REVIEW

File: gift_card.py [hunk 1]
```diff
@@ -348,3 +352,129 @@ def parse(cls, report, records, data, localcontext):
...
+    __name__ = 'gift_card.redeem.start'
+
+    description = fields.Text('Description', required=True)
+    gateway = fields.Many2One(
+        'payment_gateway.gateway', 'Gateway', required=True,
+        domain=[
+            ('method', '=', 'gift_card'),
+        ]
+    )
+    party = fields.Many2One('party.party', 'Recipient', required=True)
```
Comment (aroraumang) [inline gift_card.py]: Title should be Party only

File: gift_card.py [hunk 2]
```diff
@@ -348,3 +352,129 @@ def parse(cls, report, records, data, localcontext):
...
+    __name__ = 'gift_card.redeem.end'
+
+    done_msg = fields.Text("Message", readonly=True)
+
+    @staticmethod
+    def default_done_msg():  # pragma: no cover
+        """
+        Returns default message.
+        """
+        return (
```
Comment (aroraumang) [inline gift_card.py]: I think this could come in 1 single line

File: gift_card.py [hunk 3]
```diff
@@ -348,3 +352,129 @@ def parse(cls, report, records, data, localcontext):
...
+    "Gift Card Redeem Wizard"
+    __name__ = 'gift_card.redeem.wizard'
+
+    start = StateView(
+        'gift_card.redeem.start',
+        'gift_card.redeem_start_view_form',
+        [
+            Button('Cancel', 'end', 'tryton-cancel'),
+            Button(
+                'Save', 'do',
```
Comment (aroraumang) [inline gift_card.py]: Redeem, redeem

File: gift_card.py [hunk 4]
```diff
@@ -348,3 +352,129 @@ def parse(cls, report, records, data, localcontext):
...
+                'tryton-ok', default=True
+            )
+        ]
+    )
+    do = StateTransition()
```
Comment (aroraumang) [inline gift_card.py]: redeem

File: gift_card.py [hunk 5]
```diff
@@ -348,3 +352,129 @@ def parse(cls, report, records, data, localcontext):
...
+        })
+
+    def default_start(self, data):
+        """
+        Initial state of redeem wizard.
+        """
+        Gateway = Pool().get('payment_gateway.gateway')
+        GiftCard = Pool().get('gift_card.gift_card')
+
+        gift_card = GiftCard(Transaction().context.get('active_id'))
```
Comment (aroraumang) [inline gift_card.py]: What if the user runs the wizard for multiple records?

File: gift_card.py [hunk 6]
```diff
@@ -348,3 +352,129 @@ def parse(cls, report, records, data, localcontext):
...
+        self.check_giftcard_state(gift_card)
+
+        gateways = Gateway.search([
+            ('method', '=', 'gift_card'),
+        ])
+        return {
+            'description': '',
```
Comment (aroraumang) [inline gift_card.py]: Only return gateway

File: gift_card.py [hunk 7]
```diff
@@ -348,3 +352,129 @@ def parse(cls, report, records, data, localcontext):
...
+    def transition_do(self):
+        """
+        Transition where PaymentTransaction is created and associated
+        with current gift card.
+        """
+        GiftCard = Pool().get('gift_card.gift_card')
+        PaymentTransaction = Pool().get('payment_gateway.transaction')
+        Date = Pool().get('ir.date')
+
+        gift_card = GiftCard(Transaction().context.get('active_id'))
```
Comment (aroraumang) [inline gift_card.py]: What if the user runs the wizard for multiple records?

File: gift_card.py [hunk 8]
```diff
@@ -348,3 +352,129 @@ def parse(cls, report, records, data, localcontext):
...
+
+        transaction, = PaymentTransaction.create([{
+            'description': self.start.description or 'Redeeming gift card',
```
Comment (aroraumang) [inline gift_card.py]: No need of or statement as the field is required

File: gift_card.py [hunk 9]
```diff
@@ -348,3 +352,129 @@ def parse(cls, report, records, data, localcontext):
         return super(GiftCardReport, cls).parse(
             report, records, data, localcontext
         )
+
+
+class GiftCardRedeemStart(ModelView):
+    "Gift Card Redeem Start View"
+    __name__ = 'gift_card.redeem.start'
+
```
Comment (aroraumang) [inline gift_card.py]: The user must also be given the option to enter the amount. Also check if the amount to be redeemed is less than the available amount in GC

File: gift_card.py [hunk 10]
```diff
@@ -348,3 +352,129 @@ def parse(cls, report, records, data, localcontext):
...
+        transaction, = PaymentTransaction.create([{
+            'description': self.start.description or 'Redeeming gift card',
+            'date': Date.today(),
+            'party': self.start.party.id,
+            'address': (
+                self.start.party.addresses and
+                self.start.party.addresses[0].id or None
+            ),
+            'amount': gift_card.amount,
```
Comment (aroraumang) [inline gift_card.py]: The amount entered by user

File: gift_card.py [hunk 11]
```diff
@@ -348,3 +352,129 @@ def parse(cls, report, records, data, localcontext):
...
+            'currency': gift_card.currency.id,
+            'gateway': self.start.gateway.id,
+            'gift_card': gift_card.id,
+        }])
+        GiftCard.write([gift_card], {
+            'payment_transactions': [('add', [transaction.id])],
+        })
+        transaction.capture_self()
+        transaction.post(gift_card.payment_transactions)
```
Comment (aroraumang) [inline gift_card.py]: Just capture. It must automatically post https://github.com/openlabs/payment-gateway/blob/develop/manual.py#L70

File: gift_card.py (hunk 9 above)
Comment (PritishC) [inline gift_card.py]: Done.

File: gift_card.py [hunk 12]
```diff
@@ -348,3 +352,146 @@ def parse(cls, report, records, data, localcontext):
...
+    )
+    party = fields.Many2One('party.party', 'Party', required=True)
+    amount = fields.Numeric('Amount', digits=(16, 2), required=True)
+
+
+class GiftCardRedeemDone(ModelView):
+    "Gift Card Redeem Done View"
+    __name__ = 'gift_card.redeem.end'
+
+    done_msg = fields.Text("Message", readonly=True)
```
Comment (aroraumang) [inline gift_card.py]: No need to add a field. Just add a label in view.
Refer this https://github.com/tryton/trytond/blob/develop/trytond/ir/view/module_install_upgrade_done_form.xml#L6

File: gift_card.py [hunk 13]
```diff
@@ -348,3 +352,146 @@ def parse(cls, report, records, data, localcontext):
...
+
+    @classmethod
+    def __setup__(cls):
+        super(GiftCardRedeemWizard, cls).__setup__()
+        cls._error_messages.update({
+            'gift_card_inactive': (
+                'The gift card to be redeemed must be in active state.'
+            ),
+            'gift_card_redeemed': (
+                'This gift card has already been redeemed once.'
```
Comment (aroraumang) [inline gift_card.py]: This gift card has already been redeemed.

File: gift_card.py [hunk 14]
```diff
@@ -348,3 +352,146 @@ def parse(cls, report, records, data, localcontext):
...
+            ),
+            'multiple_giftcards': 'Please select a single gift card.',
```
Comment (aroraumang) [inline gift_card.py]: You can redeem only 1 gift card at a time.

File: gift_card.py [hunk 15]
```diff
@@ -348,3 +352,146 @@ def parse(cls, report, records, data, localcontext):
...
+            'amount_exceeded': (
+                'The amount entered was greater than amount available in the'
```
Comment (aroraumang) [inline gift_card.py]: * is greater than

File: gift_card.py [hunk 16]
```diff
@@ -348,3 +352,146 @@ def parse(cls, report, records, data, localcontext):
...
+        GiftCard = Pool().get('gift_card.gift_card')
+
+        gift_cards = GiftCard.browse(Transaction().context.get('active_ids'))
+        if len(gift_cards) > 1:
+            self.raise_user_error('multiple_giftcards')
+        gift_card = gift_cards[0]
+        self.check_giftcard_state(gift_card)
+
+        gateways = Gateway.search([
+            ('method', '=', 'gift_card'),
```
Comment (aroraumang) [inline gift_card.py]: ('active', '=', True)

File: gift_card.py [hunk 17]
```diff
@@ -348,3 +352,146 @@ def parse(cls, report, records, data, localcontext):
...
+            ),
+            'amount': self.start.amount,
+            'currency': self.start.gift_card.currency.id,
+            'gateway': self.start.gateway.id,
+            'gift_card': self.start.gift_card.id,
+        }])
+        GiftCard.write([self.start.gift_card], {
+            'payment_transactions': [('add', [transaction.id])],
+        })
+        transaction.capture_self()
```
Comment (aroraumang) [inline gift_card.py]: PaymentTransaction.capture([transaction])

File: view/redeem_done_form.xml [hunk 18]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0"?>
+<form string="Redemption Complete">
+	<field name="done_msg"/>
```
Comment (aroraumang) [inline view/redeem_done_form.xml]: Use label instead. Explained above

File: gift_card.py (hunk 16 above)
Comment (PritishC) [inline gift_card.py]: Done.

**Comment by aroraumang** [comment]
LGTM

File: gift_card.py [hunk 19]
```diff
@@ -348,3 +352,141 @@ def parse(cls, report, records, data, localcontext):
...
+            self.raise_user_error('gift_card_redeemed')
+        elif gift_card.state != 'active':
+            self.raise_user_error('gift_card_inactive')
+
+    def validate_amount(self, amount, gift_card):
+        """
+        Checks whether the amount entered by the user is not greater than the
+        amount available in the gift card.
+        """
+        if amount > gift_card.amount_available:
```
Comment (tarunbhardwaj) [inline gift_card.py]: check for -ve amount also

File: gift_card.py [hunk 20]
```diff
@@ -348,3 +352,141 @@ def parse(cls, report, records, data, localcontext):
...
+
+        self.validate_amount(self.start.amount, self.start.gift_card)
+
+        transaction, = PaymentTransaction.create([{
+            'description': self.start.description,
+            'date': Date.today(),
+            'party': self.start.party.id,
+            'address': (
+                self.start.party.addresses and
+                self.start.party.addresses[0].id or None
```
Comment (tarunbhardwaj) [inline gift_card.py]: Don't use index 0 address. ask user for this and there you can set default to invoice address or index 0

File: view/redeem_done_form.xml (hunk 18 above)
Comment (PritishC) [inline view/redeem_done_form.xml]: Changed on @tarunbhardwaj's suggestion.

File: gift_card.py (hunk 19 above)
Comment (PritishC) [inline gift_card.py]: Done.

**Comment by PritishC** [comment]
selection_051

Also added a message for the end view.

File: __init__.py [hunk 21]
```diff
@@ -6,7 +6,10 @@
     :license: BSD, see LICENSE for more details.
 """
 from trytond.pool import Pool
-from .gift_card import GiftCard, GiftCardReport
+from .gift_card import (
```
Comment (prakashpp) [inline __init__.py]: no relative import

File: gift_card.py [hunk 22]
```diff
@@ -348,3 +352,192 @@ def parse(cls, report, records, data, localcontext):
...
+        * If the party has no invoice addresses but has addresses, the first
+          among those addresses is shown.
+        * If the party has no addresses, the user gets to select them.
+        """
+        Address = Pool().get('party.address')
+
+        if self.party is None:  # If the party is removed altogether
+            return None
+
+        addresses = Address.search([
```
Comment (prakashpp) [inline gift_card.py]:
```
try:
    address, = Address.search([('party', '=', 'self.party.id')], order_by=[('invoice', 'ASC')], limit=1)
excpet ValueError:
    return None
return address.id
```

File: gift_card.py [hunk 23]
```diff
@@ -348,3 +352,192 @@ def parse(cls, report, records, data, localcontext):
...
+            'description': self.start.description,
+            'date': Date.today(),
+            'party': self.start.party.id,
+            'address': self.start.address,
+            'amount': self.start.amount,
+            'currency': self.start.gift_card.currency.id,
+            'gateway': self.start.gateway.id,
+            'gift_card': self.start.gift_card.id,
+        }])
+        GiftCard.write([self.start.gift_card], {
```
Comment (prakashpp) [inline gift_card.py]: why again?

File: gift_card.py [hunk 24]
```diff
@@ -348,3 +352,192 @@ def parse(cls, report, records, data, localcontext):
...
+    def transition_redeem(self):
+        """
+        Transition where PaymentTransaction is created and associated
+        with current gift card.
+        """
+        GiftCard = Pool().get('gift_card.gift_card')
+        PaymentTransaction = Pool().get('payment_gateway.transaction')
+        Date = Pool().get('ir.date')
+
+        self.validate_amount(self.start.amount, self.start.gift_card)
```
Comment (prakashpp) [inline gift_card.py]: remove. This is already checked in gateway code

File: gift_card.py [hunk 25]
```diff
@@ -348,3 +352,166 @@ def parse(cls, report, records, data, localcontext):
...
+        'payment_gateway.gateway', 'Gateway', required=True,
+        domain=[
+            ('method', '=', 'gift_card'),
+        ]
+    )
+    gift_card = fields.Many2One(
+        'gift_card.gift_card', 'Gift Card', readonly=True
+    )
+    party = fields.Many2One('party.party', 'Party', required=True)
+    amount = fields.Numeric('Amount', digits=(16, 2), required=True)
```
Comment (sharoonthomas) [inline gift_card.py]: add currency and the digits should come in from currency digits

File: gift_card.py [hunk 26]
```diff
@@ -348,3 +352,166 @@ def parse(cls, report, records, data, localcontext):
...
+    address = fields.Many2One(
```
Comment (sharoonthomas) [inline gift_card.py]: not needed

**Comment by prakashpp** [comment]
LGTM

File: gift_card.py [hunk 27]
```diff
@@ -348,3 +352,166 @@ def parse(cls, report, records, data, localcontext):
...
+        'party.address', 'Billing Address', required=True,
+        domain=[('party', '=', Eval('party'))],
+        depends=['party']
+    )
+
+    @fields.depends('party')
+    def on_change_with_address(self, name=None):  # pragma: no cover
```
Comment (sharoonthomas) [inline gift_card.py]: not required

File: gift_card.py [hunk 28]
```diff
@@ -348,3 +352,166 @@ def parse(cls, report, records, data, localcontext):
...
+        return 'done'
+
+    def default_done(self, data):
+        """
+        Returns a message with relevant details.
+        """
+        currency = self.start.gift_card.currency.code
+        return {
+            'done_msg': 'The gift card was redeemed with {0} {3} captured in'
+            ' this transaction, {1} {3} captured overall and {2} {3} remaining.'
```
Comment (sharoonthomas) [inline gift_card.py]: liked this :)

File: gift_card.py [hunk 29]
```diff
@@ -348,3 +352,166 @@ def parse(cls, report, records, data, localcontext):
...
+        PaymentTransaction = Pool().get('payment_gateway.transaction')
+        Date = Pool().get('ir.date')
+
+        transaction, = PaymentTransaction.create([{
+            'description': self.start.description,
+            'date': Date.today(),
+            'party': self.start.party.id,
+            'address': self.start.address,
+            'amount': self.start.amount,
+            'currency': self.start.gift_card.currency.id,
```
Comment (sharoonthomas) [inline gift_card.py]: nope, get currency from form

File: gift_card.py [hunk 30]
```diff
@@ -348,3 +352,166 @@ def parse(cls, report, records, data, localcontext):
...
             report, records, data, localcontext
         )
+
+
+class GiftCardRedeemStart(ModelView):
+    "Gift Card Redeem Start View"
+    __name__ = 'gift_card.redeem.start'
+
+    description = fields.Text('Description', required=True)
+    gateway = fields.Many2One(
```
Comment (sharoonthomas) [inline gift_card.py]: If there is only 1 gateway of the method, set it as default value

File: gift_card.py [hunk 31]
```diff
@@ -348,3 +352,166 @@ def parse(cls, report, records, data, localcontext):
...
+        })
+
+    def default_start(self, data):
+        """
+        Initial state of redeem wizard.
+        """
+        Gateway = Pool().get('payment_gateway.gateway')
+        GiftCard = Pool().get('gift_card.gift_card')
+
+        gift_cards = GiftCard.browse(Transaction().context.get('active_ids'))
```
Comment (sharoonthomas) [inline gift_card.py]: before browse check len of active_ids. Browse is expensive

File: gift_card.py (hunk 25 above)
Comment (PritishC) [inline gift_card.py]: Done.

File: gift_card.py (hunk 30 above)
Comment (PritishC) [inline gift_card.py]: Done.

File: gift_card.py (hunk 28 above)
Comment (PritishC) [inline gift_card.py]: Thanks!

**Pull request merged**